File: godot/scripts/orbital/KeplerianElements.cs
// Compliance Note
// This project is based entirely on publicly available academic information and general knowledge of orbital mechanics. It contains no restricted, proprietary, or export-controlled information of any kind. This is a personal learning project only.

namespace DeltaWrecker.Orbital;

/// <summary>
/// Represents the Keplerian orbital elements that define an orbit.
/// Mirrors engine/src/orbital/keplerian_elements.rs.
///
/// For Phase 1 only the semi-major axis is required; all other elements are
/// nullable so the data structure stays minimal while remaining extensible.
/// </summary>
public sealed class KeplerianElements
{
    /// <summary>
    /// Semi-major axis of the orbit in kilometers (km).
    /// Average distance from the central body to the orbiting object.
    /// </summary>
    public double SemiMajorAxis { get; }

    /// <summary>
    /// Eccentricity of the orbit (dimensionless).
    /// 0.0 = perfectly circular, &lt; 1.0 = elliptical.
    /// Must be &lt; 0.05 for a near-circular orbit (Phase 1).
    /// </summary>
    public double? Eccentricity { get; init; }

    /// <summary>Inclination relative to the reference plane in radians.</summary>
    public double? Inclination { get; init; }

    /// <summary>Longitude of the ascending node in radians.</summary>
    public double? LongitudeOfAscendingNode { get; init; }

    /// <summary>Argument of periapsis in radians.</summary>
    public double? ArgumentOfPeriapsis { get; init; }

    /// <summary>Mean anomaly at a reference time in radians.</summary>
    public double? MeanAnomaly { get; init; }

    /// <param name="semiMajorAxis">Semi-major axis in km. All other elements default to null.</param>
    public KeplerianElements(double semiMajorAxis)
    {
        SemiMajorAxis = semiMajorAxis;
    }

    /// <summary>Returns true if the orbit is near-circular (eccentricity &lt; 0.05).</summary>
    public bool IsNearCircular() => (Eccentricity ?? 0.0) < 0.05;
}

File: engine/src/orbital/keplerian_elements.rs
// **Compliance Note**
// This project is based entirely on publicly available academic information and general knowledge of orbital mechanics. It contains no restricted, proprietary, or export-controlled information of any kind. This is a personal learning project only.

/// Represents the Keplerian orbital elements that define an orbit.
///
/// Keplerian elements provide a compact representation of an orbit's shape,
/// orientation, and position. Currently only the semi-major axis is required,
/// with other elements available as options for future implementation.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct KeplerianElements {
    /// Semi-major axis of the orbit in kilometers (km)
    /// This is the average distance from the central body to the orbiting object
    pub semi_major_axis: f64,
    /// Eccentricity of the orbit (dimensionless, 0 = circular, <1 = elliptical)
    pub eccentricity: Option<f64>,
    /// Inclination of the orbit relative to the reference plane in radians
    pub inclination: Option<f64>,
    /// Longitude of the ascending node in radians
    pub longitude_of_ascending_node: Option<f64>,
    /// Argument of periapsis in radians
    pub argument_of_periapsis: Option<f64>,
    /// Mean anomaly at a reference time in radians
    pub mean_anomaly: Option<f64>,
}

impl KeplerianElements {
    /// Creates a new KeplerianElements with the required semi-major axis.
    ///
    /// Other orbital elements can be set to None for future use.
    ///
    /// # Arguments
    ///
    /// * `semi_major_axis` - The semi-major axis in km
    ///
    /// # Example
    ///
    /// ```
    /// use delta_wrecker::orbital::KeplerianElements;
    /// let elements = KeplerianElements::new(7000.0);
    /// assert_eq!(elements.semi_major_axis, 7000.0);
    /// ```
    pub fn new(semi_major_axis: f64) -> Self {
        KeplerianElements {
            semi_major_axis,
            eccentricity: None,
            inclination: None,
            longitude_of_ascending_node: None,
            argument_of_periapsis: None,
            mean_anomaly: None,
        }
    }

    /// Validates that the orbit is near-circular (e < 0.05).
    /// Returns true if valid, false otherwise.
    pub fn is_near_circular(&self) -> bool {
        // TODO-Fred: Implement near-circular validation logic (e < 0.05)
        // Remember to handle the Option<f64> correctly.
        unimplemented!()
    }
}

// **Compliance Note**
// This project is based entirely on publicly available academic information and general knowledge of orbital mechanics. It contains no restricted, proprietary, or export-controlled information of any kind. This is a personal learning project only.